src/v/transform/txn_reader.cc

```diff
@@ -0,0 +1,219 @@
+    }
+
+    drain_result end_of_stream() { return std::move(_result); }
+
+private:
+    void process(model::record_batch batch) {
+        model::record_batch_attributes attrs = batch.header().attrs;
+        // It's possible that there are only control batches, which then we
+        // don't have to filter out any aborts.
+        if (attrs.is_transactional()) {
+            if (!_result.txn_range) {
+                _result.txn_range.emplace(
+                  batch.base_offset(), batch.last_offset());
+            }
+            _result.txn_range->second = batch.last_offset();
+        }
+        _result.batches.push_back(std::move(batch));
+    }
+
+    drain_result _result;
+};
+
+bool is_aborted(
+  const model::record_batch& batch,
+  const absl::btree_set<model::tx_range, std::greater<>>& aborted_txns) {
+    model::producer_identity pid = {
+      batch.header().producer_id, batch.header().producer_epoch};
+    auto it = aborted_txns.lower_bound(
+      {pid, batch.base_offset(), model::offset::max()});
+
+    if (it == aborted_txns.end()) {
+        return false;
+    }
+    // If the pids match, and the batch lies within the abort range, then this
+    // batch is aborted.
+    return it->pid == pid && batch.base_offset() >= it->first
+           && batch.last_offset() <= it->last;
+}
+
+std::unique_ptr<model::record_batch_reader::impl> make_txn_filtered_reader(
+  ss::chunked_fifo<model::record_batch> batches,
+  const absl::btree_set<model::tx_range, std::greater<>>& aborted_txns) {
+    ss::chunked_fifo<model::record_batch> filtered;
+    for (auto& batch : batches) {
+        auto attrs = batch.header().attrs;
+        if (attrs.is_transactional() && is_aborted(batch, aborted_txns)) {
+            continue;
+        }
+        if (attrs.is_control()) {
+            continue;
+        }
+        filtered.push_back(std::move(batch));
+    }
+    return model::make_fragmented_memory_record_batch_reader(
+             std::move(filtered))
+      .release();
+}
+
+} // namespace
+
+read_committed_reader::read_committed_reader(
+  std::unique_ptr<aborted_transaction_tracker> tracker,
+  model::record_batch_reader reader)
+  : _tracker(std::move(tracker))
+  , _underlying(std::move(reader).release()) {}
+
+void read_committed_reader::print(std::ostream& os) {
+    os << "transform::txn_reader{}";
+}
+
+bool read_committed_reader::is_end_of_stream() const {
+    return _underlying->is_end_of_stream();
+}
+
+ss::future<model::record_batch_reader::storage_t>
+read_committed_reader::do_load_slice(
+  model::timeout_clock::time_point deadline) {
+    model::record_batch_reader::storage_t loaded
+      = co_await _underlying->do_load_slice(deadline);
+    // We delete the tracker when we've filtered the stream already.
+    if (!_tracker || !contains_control_or_txn_batch(loaded)) {
+        co_return loaded;
+    }
+    auto [txn_offset_range, batches] = co_await _underlying->consume(
+      drainer(std::move(loaded)), deadline);
```

Comment:
IIUC the only purpose of drainer is to compute begin / end offsets of the batch stream, wondering if we can merge this pass with contains_control_or_txn_batch(), avoids an extra pass when there are transactional batches.

Reply:
This is slightly simpler IMO, which is what I was going for, as I assume the extra loop over a single slice wouldn't be expensive enough to justify a more complex implementation. Let me take a stab at this tho

Reply:
Yeah the code that:

* passes through slices that don't contain any control/txn batches
* if it encounters control/txn batches switches modes and adds the already looped over batches to a new buffer and then looks for the txn offsets in the remaining batches.

Is much more complex than what I have now, and additionally we only get the double pass for a single slice so until I do some benchmarking and see this be an actual performance issue, I'm going to keep the simpler code.